### PR TITLE
Add MSVC support

### PR DESCRIPTION
**Pull request by Mixaill**
This patch set contains various improvements which allows to build DXVK on MSVC compiler.

To build DXVK you also need:

* apply this patch to meson https://github.com/mesonbuild/meson/pull/2836
* get MSVC-compatible SPIRV-Tools (you can use https://github.com/microsoft/vcpkg with this PR https://github.com/Microsoft/vcpkg/pull/2959 ) and place them into lib/lib32 folder

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -1,8 +1,12 @@
-project('dxvk', ['c', 'cpp'], default_options : ['cpp_std=c++1z'])
+project('dxvk', ['c', 'cpp'], default_options : ['cpp_std=c++17'])
```
Comment (qwertychouskie) [inline meson.build]: This should probably only be changed for MSVC as changing it for GCC could cause some older versions to not compile it even though they can.

File: src/dxvk/dxvk_memory.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ namespace dxvk {
   DxvkMemory& DxvkMemory::operator = (DxvkMemory&& other) {
     m_chunk   = std::exchange(other.m_chunk,  nullptr);
     m_heap    = std::exchange(other.m_heap,   nullptr);
-    m_memory  = std::exchange(other.m_memory, VkDeviceMemory(nullptr));
+    m_memory  = std::exchange(other.m_memory, VkDeviceMemory(0));
```
Comment (doitsujin) [inline src/dxvk/dxvk_memory.cpp]: this should be `VK_NULL_HANDLE` instead of 0 or `nullptr`.

Comment (doitsujin) [reply src/dxvk/dxvk_memory.cpp]: Settng the C++ version to c++17 will break GCC 6.x, which is what currently ships with Ubuntu. Breaking that is questionable, but I'm not sure if there is a way to set different values for different compilers.

File: src/d3d11/d3d11_context.h [hunk 3]
```diff
@@ -7,6 +7,10 @@
 #include "d3d11_context_state.h"
 #include "d3d11_device_child.h"
 
+#ifndef WINBOOL
+typedef BOOL WINBOOL;
```
Comment (doitsujin) [inline src/d3d11/d3d11_context.h]: I think `WINBOOL` shouldn't be a thing in the first place, if it's the same as `BOOL` then I think we should rather get rid of all uses of `WINBOOL` instead.

**Comment by Mixaill** [comment]
*  `WINBOOL` replaced with `BOOL`
* 0 replaced with `VK_NULL_HANDLE`
* `cpp_std=c++17` only for MSVC.

**Comment by doitsujin** [comment]
Had to revert the changes because it breaks mingw builds. mingw does not support `.def` files and exits with an error message when linking the libraries. Please open a new PR once that is sorted out.

**Comment by pchome** [comment]
@Mixaill 
I see this changes mostly echoed with those one from #60 . Now I also using `.def` files and redefined `meson.build` files but in a bit different way ( https://github.com/pchome/dxvk ).

Last changes was to create `./linux/` directory and move every specific change there to less interfere w/ upstream changes. It mostly WIP and POC but maybe some ideas would be useful for you.

And maybe you'll find even better way to merge both linux and MSVC builds together.

**Comment by pchome** [comment]
@Mixaill 
Also there is ability to create custom meson build file set and patches with [wrap feature](http://mesonbuild.com/Wrap-dependency-system-manual.html). As last chance to extend some feature.

**Comment by pchome** [comment]
BTW: `subdir()` it's mesons `include` - includes everything from specified dir in place it defined.
So possible to do something like this: https://github.com/doitsujin/dxvk/commit/b771995958e2f8a942872495349279c043e17e4c

**Comment by Mixaill** [comment]
I wonder, if CMake will be easier for that crossplatform crosscompiler mess?

**Comment by doitsujin** [comment]
One thing just came to mind, can you try compiling DXVK on MSVC++ with the MinGW D3D11 headers instead of the system headers? Maybe that way we could avoid using `.def` files. As far as I'm aware, the main issue is that the Windows headers are meant for application developers who *import* the D3D libs, not for crazy Linux people trying to re-implement the entire API.

**Comment by Mixaill** [comment]
>can you try compiling DXVK on MSVC++ with the MinGW D3D11 headers instead of the Windows SDK headers? 

It will require some nontrivial black magic.

-----

BTW, my MinGW form MSYS2 project has all the `D3D11_` header and doesn't have _countof macro =\

**Pull request merged**